src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,887 @@
+
+	d_iov_t      iov;
+	d_sg_list_t  sgl;
+
+	/* Buffer view that implements Python Buffer Protocol, has to be release after use */
+	Py_buffer    buf_view;
+};
+
+static int
+start_read_op(struct dfs_handle *hdl, PyObject *item, struct io_op *op)
+{
+	assert(op != NULL);
+
+	int           rc  = 0;
+	daos_event_t *evp = &op->ev;
+
+	PyObject     *py_path = PyTuple_GetItem(item, 0);
+	PyObject     *py_buff = PyTuple_GetItem(item, 1);
+
+	if (py_path == NULL || py_buff == NULL) {
+		D_WARN("Each tuple must contain exactly two elements: path and bytearray");
+		return EINVAL;
+	}
+
+	const char *path = PyUnicode_AsUTF8(py_path);
+	if (path == NULL) {
+		D_WARN("First element of a tuple does not look like a path");
+		return EINVAL;
+	}
+
+	if (PyObject_GetBuffer(py_buff, &op->buf_view, PyBUF_WRITE) == -1) {
+		D_WARN("Buffer is not writable");
+		return EINVAL;
+	}
+
+	if (!PyBuffer_IsContiguous(&op->buf_view, 'C')) {
+		D_WARN("Buffer for '%s' is not contiguous", path);
+		rc = EINVAL;
+		goto out;
+	}
+
+	rc = daos_event_init(evp, hdl->eq, NULL);
+	if (rc) {
+		D_ERROR("Could not init event: %s (rc=%d)", d_errstr(rc), rc);
+		goto out;
+	}
+
+	rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &op->obj, NULL, NULL);
+	if (rc) {
+		D_ERROR("Could not lookup path '%s': %s (rc=%d)", op->path, strerror(rc), rc);
+		goto out;
+	}
+
+	op->path = path;
+	op->size = op->buf_view.len;
+	d_iov_set(&op->iov, op->buf_view.buf, op->size);
+
+	op->sgl.sg_nr     = 1;
+	op->sgl.sg_nr_out = 0;
+	op->sgl.sg_iovs   = &op->iov;
+
+	rc = dfs_read(hdl->dfs, op->obj, &op->sgl, 0 /* offset */, &op->size, &op->ev);
+	if (rc) {
+		D_ERROR("Could not start async read on '%s': %s (rc=%d)", op->path, strerror(rc),
+			rc);
+		goto out;
+	}
+
+	return DER_SUCCESS;
+
+out:
+	PyBuffer_Release(&op->buf_view);
+
+	int rc2 = daos_event_fini(&op->ev);
+	if (rc2) {
+		D_ERROR("Could not finalize event: %s (rc=%d)", d_errstr(rc2), rc2);
+	}
+
+	if (op->obj != NULL) {
+		rc2 = dfs_release(op->obj);
+		if (rc2) {
+			D_ERROR("Could not release obj handler of '%s': %s (rc=%d)", op->path,
+				strerror(rc2), rc2);
+		}
+		op->obj = NULL;
+	}
+
+	return rc;
+}
+
+static int
+complete_read_op(struct dfs_handle *hdl, struct io_op *op)
+{
+	assert(op != NULL);
+
+	D_DEBUG(DB_ANY, "READ of %zu bytes from '%s' completed with  status: %s (rc = %d)",
+		op->size, op->path, d_errstr(op->ev.ev_error), op->ev.ev_error);
+
+	int rc = dfs_release(op->obj);
+	if (rc) {
+		D_WARN("Could not release object handler %s: %s (rc=%d)", op->path, strerror(rc),
+		       rc);
+	}
+	op->obj = NULL;
+
+	op->err = op->ev.ev_error;
+
+	rc = daos_event_fini(&op->ev);
+	if (rc) {
+		D_WARN("Could not finalize event handler of '%s': %s (rc=%d)", op->path,
+		       d_errstr(rc), rc);
+	}
+
+	PyBuffer_Release(&op->buf_view);
+	return 0;
+}
+
+static int
+reap_read_op(struct dfs_handle *hdl, struct io_op **op)
+{
+	daos_event_t *evp = NULL;
+
+	assert(op != NULL);
+
+	int rc = daos_eq_poll(hdl->eq, 1, DAOS_EQ_WAIT, 1, &evp);
```

Comment:
why use a timeout of 1?

Reply:
do not use EQ_WAIT, it's bad and has issues on some providers.
use EQ_NOWAIT in a loop, and you can add a sleep or yield in the loop.
also it's better to poll on multiple events, not just 1.

Reply:
The second parameter for daos_eq_poll is not a timeout but the flag `wait_running` which tells `daos_eq_poll` do not wait if there are no running events.

Could you please elaborate why it's bad to use EQ_WAIT in case when my intention is to way until any event finishes in the queue on the current thread, so the next item can bee queued ?
The intention is to limit number of in-flight operations.
I'm 100% agree that polling multiple events is better but probably requires additional bookeeping. I'll have a look at that.

Reply:
In general it's bad practice to do it that way if you have multiple events in flight. it's always better to poll all events and not wait indefinitely for one.
That being said, we have seen issues with some providers in ucx that if you poll on some events but not others, you get timeouts and hangs. so we changed to use NOWAIT for that purpose to resolve such issues.

Also as I mentioned, we already have a way in DAOS to limit inflight RPCs of a process with an env variable. I do not think your should be managing that on this layer.

Reply:
@mchaarawi Sounds good, I've pushed the updated version.